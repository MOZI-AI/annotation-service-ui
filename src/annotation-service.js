import React from 'react';
import { GeneSelectionForm } from './gene-selection';
import { AnnotationSelection } from './annotation-selection'
import { Divider, Button, Row, Icon, message } from 'antd'

const availableAnnotations = [
    'Gene Ontology',
    'Reactome',
    'Small Molecules'
]

export class AnnotationService extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            genes: [],
            geneList: null,
            selectedAnnotations: []
        }
        // bind functions
        this.handleGeneAdded = this.handleGeneAdded.bind(this)
        this.handleGeneRemoved = this.handleGeneRemoved.bind(this)
        this.handleGeneListUploaded = this.handleGeneListUploaded.bind(this)
        this.handleAllGenesRemoved = this.handleAllGenesRemoved.bind(this)
        this.handleAnnotationsChanged = this.handleAnnotationsChanged.bind(this)
        this.handleAnnotationFilterChanged = this.handleAnnotationFilterChanged.bind(this)
    }


    handleGeneAdded(gene) {
        this.setState(state => {
            const genes = state.genes.slice(0);
            genes.push(gene);
            return { genes: genes }
        })
    }


    handleGeneRemoved(gene) {
        this.setState(state => {
            let genes = state.genes.slice(0);
            return { genes: genes.filter(g => g !== gene) }
        })
    }


    handleGeneListUploaded(geneList) {
        const fileReader = new FileReader();
        fileReader.readAsText(geneList);
        fileReader.onload = () => {
            // The file is supposed to contain one gene per line. Furthermore, gene names are supposed to contain alphanumeric characters only
            const re = /^[a-z0-9\s]+$/i;
            if (!re.test(fileReader.result)) {
                message.error("The selected file contains invalid characters.");
                return;
            }
            const geneArray = fileReader.result.split('\n')
            const uniqueGeneArray = geneArray.reduce((accumulator, value) => {
                if (value && accumulator.indexOf(value) === -1) accumulator.push(value);
                return accumulator;
            }, [])
            this.setState({ geneList: geneList, genes: uniqueGeneArray })
        };
    }


    handleAllGenesRemoved() {
        this.setState({ genes: [], geneList: null })
    }


    handleAnnotationsChanged(e) {
        this.setState(state => {
            let selectedAnnotations = state.selectedAnnotations.slice();
            e.target.checked ? selectedAnnotations.push({ name: e.target.name, filter: {} })
                : selectedAnnotations = selectedAnnotations.filter(a => a.name !== e.target.name);

            return { selectedAnnotations: selectedAnnotations }
        })
    }


    handleAnnotationFilterChanged(annotation, e) {
        this.setState(state => {
            const selectedAnnotations = state.selectedAnnotations.slice();
            let targetAnnotation = selectedAnnotations.find(a => a.name === annotation);
            targetAnnotation.filter[e.target.name] = e.target.value;

            return { selectedAnnotations: selectedAnnotations }
        })
    }


    isFormValid() {
        return this.state.selectedAnnotations.length && this.state.genes.length
    }


    componentDidUpdate() {
        console.log('State ', JSON.stringify(this.state, null, 2))
    }


    render() {
        return (<React.Fragment>
            <GeneSelectionForm
                genes={this.state.genes}
                geneList={this.state.geneList}
                onGeneAdded={this.handleGeneAdded}
                onGeneRemoved={this.handleGeneRemoved}
                onGeneListUploaded={this.handleGeneListUploaded}
                onAllGenesRemoved={this.handleAllGenesRemoved}
            />
            <Divider dashed />
            <AnnotationSelection
                availableAnnotations={availableAnnotations}
                onAnnotationsChanged={this.handleAnnotationsChanged}
                onAnnotationFilterChanged={this.handleAnnotationFilterChanged}
                selectedAnnotations={this.state.selectedAnnotations}
            />
            <Divider dashed />
            <Row type='flex' justify='end'>
                <Button id="submit" type='primary' disabled={!this.isFormValid()} >
                    <Icon type='check' />
                    Submit
                </Button>
            </Row>
        </React.Fragment>)
    }

}
