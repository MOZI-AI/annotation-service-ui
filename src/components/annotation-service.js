import React from "react";
import { SERVER_ADDRESS, showNotification } from "../utils";
import { Annotate } from "../proto/annotation_pb_service";
import {
  AnnotationRequest,
  Annotation,
  Gene,
  Filter
} from "../proto/annotation_pb";
import { grpc } from "grpc-web-client";
import { GeneSelectionForm } from "./gene-selection";
import { AnnotationSelection } from "./annotation-selection";
import { Button, Grid } from "@material-ui/core";
import { Check, CheckCircle } from "@material-ui/icons";

export class AnnotationService extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      genes: [],
      geneList: null,
      selectedAnnotations: [],
      annotationResult: null,
      busy: false,
      notification: null
    };
    // bind functions
    this.handleGeneAdded = this.handleGeneAdded.bind(this);
    this.handleGeneRemoved = this.handleGeneRemoved.bind(this);
    this.handleGeneListUploaded = this.handleGeneListUploaded.bind(this);
    this.handleAllGenesRemoved = this.handleAllGenesRemoved.bind(this);
    this.handleAnnotationsChanged = this.handleAnnotationsChanged.bind(this);
    this.handleFilterChanged = this.handleFilterChanged.bind(this);
    this.downloadSchemeFile = this.downloadSchemeFile.bind(this);
  }

  handleGeneAdded(input) {
    this.setState(state => {
      let genes = state.genes.slice(0);
      genes = [
        ...genes,
        ...input
          .trim()
          .toUpperCase()
          .split(" ")
      ].filter((g, i, arr) => g && arr.indexOf(g) === i);
      return { genes: genes };
    });
  }

  handleGeneRemoved(gene) {
    this.setState(state => {
      let genes = state.genes.slice(0);
      return { genes: genes.filter(g => g !== gene) };
    });
  }

  handleGeneListUploaded(geneList) {
    const fileReader = new FileReader();
    fileReader.readAsText(geneList);
    fileReader.onload = () => {
      const re = /^[a-z0-9\s]+$/i;
      if (!re.test(fileReader.result)) {
        this.setState({
          notification: {
            message: "The selected file contains invalid characters.",
            busy: false
          }
        });
        return;
      }
      const geneArray = fileReader.result.split("\n");
      const uniqueGeneArray = geneArray.reduce((accumulator, value) => {
        if (value && accumulator.indexOf(value) === -1) accumulator.push(value);
        return accumulator;
      }, []);
      this.setState({ geneList: geneList, genes: uniqueGeneArray });
    };
  }

  handleAllGenesRemoved() {
    this.setState({ genes: [], geneList: null });
  }

  handleAnnotationsChanged(isSelected, annotation) {
    this.setState(state => {
      let selectedAnnotations = state.selectedAnnotations.slice();
      isSelected
        ? selectedAnnotations.push({
            name: annotation,
            filter: this.props.availableAnnotations.find(
              a => a.key === annotation
            ).defaults
          })
        : (selectedAnnotations = selectedAnnotations.filter(
            a => a.name !== annotation
          ));

      return { selectedAnnotations: selectedAnnotations };
    });
  }

  handleFilterChanged(annotation, filter) {
    this.setState(state => {
      const selectedAnnotations = state.selectedAnnotations.map(sa => {
        if (sa.name === annotation) {
          sa.filter = Object.assign({}, sa.filter, filter);
        }
        return sa;
      });
      return { selectedAnnotations: selectedAnnotations };
    });
  }

  isFormValid() {
    let valid = true;
    valid = valid && this.state.selectedAnnotations.length;
    valid = valid && this.state.genes.length;
    // If Gene GO annotation is selected, namespace must be defined
    const GO = this.state.selectedAnnotations.find(
      a => a.name === "gene_go_annotation"
    );
    if (GO) valid = valid && GO.filter.namespace.length;
    // If Gene Pathway annotation is selected, namespace must be defined
    const Pathway = this.state.selectedAnnotations.find(
      a => a.name === "gene_pathway_annotation"
    );
    if (Pathway) {
      valid =
        valid &&
        (Pathway.filter.namespace.includes("smpdb") ||
          Pathway.filter.namespace.includes("reactome"));
    }
    return valid;
  }

  downloadSchemeFile() {
    const json = `data:application/txt, ${encodeURIComponent(
      this.state.annotationResult.schemeFile
    )}`;
    const link = document.createElement("a");
    link.setAttribute("href", json);
    link.setAttribute("download", "annotations-scheme.scm");
    link.click();
  }

  capitalizeFirstLetter(string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
  }

  handleSubmit() {
    const annotationResult = new AnnotationRequest();
    annotationResult.setGenesList(
      this.state.genes.map(g => {
        const gene = new Gene();
        gene.setGenename(g);
        return gene;
      })
    );
    annotationResult.setAnnotationsList(
      this.state.selectedAnnotations.map(sa => {
        const annotation = new Annotation();
        annotation.setFunctionname(sa.name);
        annotation.setFiltersList(
          sa.filter
            ? Object.keys(sa.filter).map(k => {
                const filter = new Filter();
                filter.setFilter(k);
                filter.setValue(
                  Array.isArray(sa.filter[k])
                    ? sa.filter[k]
                        .reduce((acc, value) => {
                          return acc + " " + value;
                        }, "")
                        .trim()
                    : this.capitalizeFirstLetter(sa.filter[k].toString())
                );
                return filter;
              })
            : []
        );
        return annotation;
      })
    );

    this.setState({
      notification: { message: "Fetching annotation results ...", busy: true }
    });

    grpc.unary(Annotate.Annotate, {
      request: annotationResult,
      host: SERVER_ADDRESS,
      onEnd: ({ status, statusMessage, message }) => {
        if (status === grpc.Code.OK) {
          this.setState(state => ({
            busy: false,
            annotationResult: message.array[0],
            notification: null
          }));
        } else {
          if (statusMessage.includes("Gene Doesn't exist")) {
            const invalidGenes = statusMessage
              .split("`")[1]
              .split(",")
              .map(g => g.trim())
              .filter(g => g);
            this.setState(state => ({
              busy: false,
              genes: state.genes.filter(g => !invalidGenes.includes(g)),
              notification: { message: statusMessage, busy: false }
            }));
          } else {
            this.setState(state => ({
              busy: false,
              notification: { message: statusMessage, busy: false }
            }));
          }
        }
      }
    });
  }

  render() {
    return (
      <React.Fragment>
        {!this.state.annotationResult && (
          <div>
            {this.state.notification &&
              showNotification(this.state.notification, () => {
                this.setState({ notification: null });
              })}
            <GeneSelectionForm
              genes={this.state.genes}
              geneList={this.state.geneList}
              onGeneAdded={this.handleGeneAdded}
              onGeneRemoved={this.handleGeneRemoved}
              onGeneListUploaded={this.handleGeneListUploaded}
              onAllGenesRemoved={this.handleAllGenesRemoved}
            />
            <AnnotationSelection
              handleAnnotationsChanged={this.handleAnnotationsChanged}
              handleFilterChanged={this.handleFilterChanged}
              selectedAnnotations={this.state.selectedAnnotations}
              availableAnnotations={this.props.availableAnnotations}
            />
            <Grid container justify="flex-end">
              <Grid item>
                <Button
                  id="submitButton"
                  variant="contained"
                  onClick={() => this.handleSubmit()}
                  disabled={!this.isFormValid()}
                  color="primary"
                >
                  <Check />
                  Submit
                </Button>
              </Grid>
            </Grid>
          </div>
        )}
        {this.state.annotationResult ? (
          <Grid container justify="center">
            <Grid
              item
              style={{
                textAlign: "center",
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
                justifyContent: "center",
                height: "60vh"
              }}
            >
              <CheckCircle style={{ fontSize: "72px", color: "#54C21F" }} />
              <h1 style={{ margin: 5 }}>Your request is being processed</h1>
              <h3 style={{ marginTop: 0, color: "#555" }}>
                Follow the link below to view results.
              </h3>
              <a
                rel="noopener noreferrer"
                target="_blank"
                href={this.state.annotationResult}
                style={{ fontSize: 18 }}
              >
                {this.state.annotationResult}
              </a>
            </Grid>
          </Grid>
        ) : null}
      </React.Fragment>
    );
  }
}
