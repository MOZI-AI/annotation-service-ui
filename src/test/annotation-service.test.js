import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { AnnotationService } from '../annotation-service'
import renderer from 'react-test-renderer';

configure({ adapter: new Adapter() });


describe('<AnnotationService />', () => {

    it('renders correctly', () => {
        const tree = renderer.create(<AnnotationService />).toJSON()
        expect(tree).toMatchSnapshot()
    });


    it('renders a disabled submit button on mount', () => {
        const wrapper = shallow(<AnnotationService />)
        expect(wrapper.find('Button#submit[disabled=true]').exists()).toBeTruthy()
    })


    it('enables submit button once genes and annotations are selected', () => {
        const wrapper = shallow(<AnnotationService />)
        expect(wrapper.find('Button#submit[disabled=true]').exists()).toBeTruthy()
        wrapper.setState({ selectedAnnotations: [{ name: 'a', filter: {} }], genes: ['ACP1'] })
        expect(wrapper.find('Button#submit[disabled=false]').exists()).toBeTruthy()
    })


})