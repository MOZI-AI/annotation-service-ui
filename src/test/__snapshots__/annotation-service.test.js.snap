// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnnotationService /> renders correctly 1`] = `
Array [
  <form
    className="ant-form ant-form-horizontal"
    onSubmit={[Function]}
  >
    <div
      className="ant-row ant-form-item"
    >
      <div
        className="ant-form-item-control-wrapper"
      >
        <div
          className="ant-form-item-control"
        >
          <span
            className="ant-form-item-children"
          >
            <div
              className="ant-radio-group ant-radio-group-outline"
              id="inputMethod"
            >
              <label
                className="ant-radio-wrapper ant-radio-wrapper-checked"
              >
                <span
                  className="ant-radio ant-radio-checked"
                  style={Object {}}
                >
                  <input
                    checked={true}
                    className="ant-radio-input"
                    disabled={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="radio"
                    value={1}
                  />
                  <span
                    className="ant-radio-inner"
                  />
                </span>
                <span>
                  Input directly
                </span>
              </label>
              <label
                className="ant-radio-wrapper"
              >
                <span
                  className="ant-radio"
                  style={Object {}}
                >
                  <input
                    checked={false}
                    className="ant-radio-input"
                    disabled={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="radio"
                    value={2}
                  />
                  <span
                    className="ant-radio-inner"
                  />
                </span>
                <span>
                  Import from file
                </span>
              </label>
            </div>
          </span>
        </div>
      </div>
    </div>
  </form>,
  <form
    className="ant-form ant-form-horizontal"
    id="directInputForm"
    onSubmit={[Function]}
  >
    <div
      className="ant-row ant-form-item"
      style={
        Object {
          "marginBottom": "0",
        }
      }
    >
      <div
        className="ant-form-item-label"
      >
        <label
          className="ant-form-item-required"
          htmlFor="gene"
          onClick={[Function]}
          title="Gene name"
        >
          Gene name
        </label>
      </div>
      <div
        className="ant-form-item-control-wrapper"
      >
        <div
          className="ant-form-item-control"
        >
          <span
            className="ant-form-item-children"
          >
            <span
              className="ant-input-affix-wrapper"
            >
              <input
                className="ant-input"
                data-__field={
                  Object {
                    "name": "gene",
                  }
                }
                data-__meta={
                  Object {
                    "name": "gene",
                    "originalProps": Object {
                      "disabled": false,
                      "onChange": [Function],
                      "onPressEnter": [Function],
                      "placeholder": "Input gene name and hit enter",
                      "prefixCls": "ant-input",
                      "suffix": <Icon
                        type="enter"
                      />,
                      "type": "text",
                    },
                    "ref": null,
                    "rules": Array [
                      Object {
                        "message": "Please input gene name!",
                        "required": true,
                      },
                      Object {
                        "validator": [Function],
                      },
                    ],
                    "trigger": "onChange",
                    "validate": Array [
                      Object {
                        "rules": Array [
                          Object {
                            "message": "Please input gene name!",
                            "required": true,
                          },
                          Object {
                            "validator": [Function],
                          },
                        ],
                        "trigger": Array [
                          "onChange",
                        ],
                      },
                    ],
                    "valuePropName": "value",
                  }
                }
                disabled={false}
                id="gene"
                onChange={[Function]}
                onKeyDown={[Function]}
                placeholder="Input gene name and hit enter"
                style={null}
                type="text"
                value=""
              />
              <span
                className="ant-input-suffix"
              >
                <i
                  className="anticon anticon-enter"
                >
                  <svg
                    aria-hidden="true"
                    className=""
                    data-icon="enter"
                    fill="currentColor"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M864 170h-60c-4.4 0-8 3.6-8 8v518H310v-73c0-6.7-7.8-10.5-13-6.3l-141.9 112a8 8 0 0 0 0 12.6l141.9 112c5.3 4.2 13 .4 13-6.3v-75h498c35.3 0 64-28.7 64-64V178c0-4.4-3.6-8-8-8z"
                    />
                  </svg>
                </i>
              </span>
            </span>
          </span>
        </div>
      </div>
    </div>
  </form>,
  <div
    style={
      Object {
        "marginTop": "10px",
      }
    }
  >
    <h4
      style={
        Object {
          "color": "#82909d",
          "marginBottom": "5px",
        }
      }
    >
       Selected genes
    </h4>
    <div
      className="ant-alert ant-alert-warning ant-alert-with-description ant-alert-no-icon"
      data-show={true}
    >
      <span
        className="ant-alert-message"
      />
      <span
        className="ant-alert-description"
      >
        The genes you input appear here. You may input gene names directly or import them from file.
      </span>
    </div>
  </div>,
  <div
    className="ant-divider ant-divider-horizontal ant-divider-dashed"
  />,
  <h4
    style={
      Object {
        "marginBottom": "15px",
      }
    }
  >
    Annotations: 
  </h4>,
  <div>
    <label
      className="ant-checkbox-wrapper"
    >
      <span
        className="ant-checkbox"
        style={Object {}}
      >
        <input
          checked={false}
          className="ant-checkbox-input"
          name="Gene Ontology"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="checkbox"
        />
        <span
          className="ant-checkbox-inner"
        />
      </span>
      <span>
         
        Gene Ontology
         
      </span>
    </label>
  </div>,
  <div>
    <label
      className="ant-checkbox-wrapper"
    >
      <span
        className="ant-checkbox"
        style={Object {}}
      >
        <input
          checked={false}
          className="ant-checkbox-input"
          name="Reactome"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="checkbox"
        />
        <span
          className="ant-checkbox-inner"
        />
      </span>
      <span>
         
        Reactome
         
      </span>
    </label>
  </div>,
  <div>
    <label
      className="ant-checkbox-wrapper"
    >
      <span
        className="ant-checkbox"
        style={Object {}}
      >
        <input
          checked={false}
          className="ant-checkbox-input"
          name="Small Molecules"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="checkbox"
        />
        <span
          className="ant-checkbox-inner"
        />
      </span>
      <span>
         
        Small Molecules
         
      </span>
    </label>
  </div>,
  <div
    className="ant-divider ant-divider-horizontal ant-divider-dashed"
  />,
  <div
    className="ant-row-flex ant-row-flex-end"
  >
    <button
      className="ant-btn ant-btn-primary"
      disabled={true}
      id="submit"
      onClick={[Function]}
      type="button"
    >
      <i
        className="anticon anticon-check"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="check"
          fill="currentColor"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M912 190h-69.9c-9.8 0-19.1 4.5-25.1 12.2L404.7 724.5 207 474a32 32 0 0 0-25.1-12.2H112c-6.7 0-10.4 7.7-6.3 12.9l273.9 347c12.8 16.2 37.4 16.2 50.3 0l488.4-618.9c4.1-5.1.4-12.8-6.3-12.8z"
          />
        </svg>
      </i>
      <span>
        Submit
      </span>
    </button>
  </div>,
]
`;
